File: android/app/src/main/java/com/vpn/android/vpn/TransportFallbackPolicy.java
package com.vpn.android.vpn;

/**
 * Phase 9 "transport flexibility": once every node has been tried on the
 * primary transport (XHTTP) without success, switch to the gRPC+Reality
 * fallback the server advertises per node (docs/PLAN.md §6: "VLESS + gRPC +
 * Reality — запасной") before giving up. Pure decision logic, driven by the
 * caller each time {@link ReconnectBackoffPolicy} signals a node switch.
 */
public class TransportFallbackPolicy {

    public enum Transport {
        XHTTP, GRPC
    }

    public static final class Outcome {
        public final boolean transportChanged;
        public final boolean allTransportsExhausted;

        Outcome(boolean transportChanged, boolean allTransportsExhausted) {
            this.transportChanged = transportChanged;
            this.allTransportsExhausted = allTransportsExhausted;
        }
    }

    private final int nodeCount;
    private final boolean grpcAvailable;
    private Transport current;
    private int switchesSinceTransportStart = 0;

    public TransportFallbackPolicy(int nodeCount, boolean grpcAvailable) {
        this(nodeCount, grpcAvailable, Transport.XHTTP);
    }

    /**
     * @param initialTransport server-advertised {@code transport_policy.primaryTransport}
     *                         (docs/PLAN.md §6/§10). Only honored when {@code grpcAvailable} —
     *                         otherwise starting on GRPC would have nowhere to fall back to.
     */
    public TransportFallbackPolicy(int nodeCount, boolean grpcAvailable, Transport initialTransport) {
        this.nodeCount = Math.max(1, nodeCount);
        this.grpcAvailable = grpcAvailable;
        this.current = (initialTransport == Transport.GRPC && grpcAvailable) ? Transport.GRPC : Transport.XHTTP;
    }

    public Transport getCurrentTransport() {
        return current;
    }

    /** Call once per node switch (i.e. whenever ReconnectBackoffPolicy.Decision.switchNode is true). */
    public Outcome onNodeSwitch() {
        switchesSinceTransportStart++;
        if (switchesSinceTransportStart < nodeCount) {
            return new Outcome(false, false);
        }

        // Every node has now been tried at least once on the current transport.
        switchesSinceTransportStart = 0;
        if (current == Transport.XHTTP && grpcAvailable) {
            current = Transport.GRPC;
            return new Outcome(true, false);
        }
        return new Outcome(false, true);
    }
}

File: desktop/src/shared/transportFallbackPolicy.ts
/**
 * Phase 9 "transport flexibility": once every node has been tried on the
 * primary transport (XHTTP) without success, switch to the gRPC+Reality
 * fallback the server advertises per node (docs/PLAN.md §6: "VLESS + gRPC +
 * Reality — запасной") before giving up. Same contract as the Android
 * client's TransportFallbackPolicy, ported independently per docs/PLAN.md §5.
 */
export type Transport = 'XHTTP' | 'GRPC';

export interface TransportSwitchOutcome {
  transportChanged: boolean;
  allTransportsExhausted: boolean;
}

export class TransportFallbackPolicy {
  private readonly nodeCount: number;
  private readonly grpcAvailable: boolean;
  private current: Transport = 'XHTTP';
  private switchesSinceTransportStart = 0;

  constructor(nodeCount: number, grpcAvailable: boolean) {
    this.nodeCount = Math.max(1, nodeCount);
    this.grpcAvailable = grpcAvailable;
  }

  getCurrentTransport(): Transport {
    return this.current;
  }

  /** Call once per node switch (i.e. whenever ReconnectBackoffPolicy's Decision.switchNode is true). */
  onNodeSwitch(): TransportSwitchOutcome {
    this.switchesSinceTransportStart += 1;
    if (this.switchesSinceTransportStart < this.nodeCount) {
      return { transportChanged: false, allTransportsExhausted: false };
    }

    this.switchesSinceTransportStart = 0;
    if (this.current === 'XHTTP' && this.grpcAvailable) {
      this.current = 'GRPC';
      return { transportChanged: true, allTransportsExhausted: false };
    }
    return { transportChanged: false, allTransportsExhausted: true };
  }
}
